File: sdks/php/src/HookRelayClient.php
<?php

declare(strict_types=1);

namespace HookRelay;

/**
 * Official PHP client for the HookRelay webhook delivery service.
 *
 * @example
 * ```php
 * $client = new HookRelayClient('hr_live_...');
 *
 * $endpoint = $client->endpoints()->create('https://myapp.com/webhook', 'Orders');
 *
 * $delivery = $client->webhooks()->send($endpoint['id'], 'order.created', [
 *     'orderId' => '12345',
 *     'amount' => 99.99,
 * ]);
 * ```
 */
class HookRelayClient
{
    private const DEFAULT_BASE_URL = 'http://localhost:3000/v1';
    private const DEFAULT_TIMEOUT = 30;

    private string $apiKey;
    private string $baseUrl;
    private int $timeout;

    private EndpointsResource $endpoints;
    private WebhooksResource $webhooks;

    public function __construct(string $apiKey, ?string $baseUrl = null, int $timeout = 0)
    {
        $this->apiKey = $apiKey;
        $this->baseUrl = rtrim($baseUrl ?? self::DEFAULT_BASE_URL, '/');
        $this->timeout = $timeout > 0 ? $timeout : self::DEFAULT_TIMEOUT;
        $this->endpoints = new EndpointsResource($this);
        $this->webhooks = new WebhooksResource($this);
    }

    public function endpoints(): EndpointsResource
    {
        return $this->endpoints;
    }

    public function webhooks(): WebhooksResource
    {
        return $this->webhooks;
    }

    /**
     * Get platform statistics.
     */
    public function getStats(): array
    {
        $resp = $this->request('GET', '/stats');
        return [
            'total_deliveries' => $resp['total_deliveries'],
            'delivered' => $resp['delivered'],
            'failed' => $resp['failed'],
            'pending' => $resp['pending'],
            'success_rate' => $resp['success_rate'],
            'endpoints_count' => $resp['endpoints_count'],
        ];
    }

    /**
     * @internal Make an API request.
     */
    public function request(string $method, string $path, ?array $body = null): mixed
    {
        $url = $this->baseUrl . $path;

        $ch = curl_init();
        curl_setopt_array($ch, [
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT => $this->timeout,
            CURLOPT_CONNECTTIMEOUT => $this->timeout,
            CURLOPT_CUSTOMREQUEST => $method,
            CURLOPT_HTTPHEADER => [
                'Authorization: Bearer ' . $this->apiKey,
                'Content-Type: application/json',
                'User-Agent: hookrelay-php/0.1.0',
            ],
        ]);

        if ($body !== null && in_array($method, ['POST', 'PUT', 'PATCH'], true)) {
            curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($body));
        }

        $response = curl_exec($ch);
        $statusCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $contentType = curl_getinfo($ch, CURLINFO_CONTENT_TYPE);
        $error = curl_error($ch);
        curl_close($ch);

        if ($response === false) {
            throw new HookRelayException("cURL error: {$error}", 0, 'NETWORK_ERROR');
        }

        if ($statusCode >= 200 && $statusCode < 300) {
            if (str_contains($contentType ?? '', 'text/csv')) {
                return $response;
            }
            return json_decode($response, true);
        }

        $message = "HTTP {$statusCode}";
        $decoded = json_decode($response, true);
        if ($decoded && isset($decoded['error']['message'])) {
            $message = $decoded['error']['message'];
        }

        return match ($statusCode) {
            400 => throw new ValidationException($message),
            401 => throw new AuthenticationException($message),
            404 => throw new NotFoundException($message),
            413 => throw new PayloadTooLargeException($message),
            429 => throw new RateLimitException($message),
            default => throw new HookRelayException($message, $statusCode, 'UNKNOWN'),
        };
    }
}

/**
 * Endpoints resource — CRUD operations for webhook endpoints.
 */
class EndpointsResource
{
    public function __construct(private HookRelayClient $client) {}

    /**
     * Create a new endpoint.
     */
    public function create(string $url, ?string $description = null, ?array $retryPolicy = null): array
    {
        $body = ['url' => $url];
        if ($description !== null) $body['description'] = $description;
        if ($retryPolicy !== null) {
            $rp = [];
            if (isset($retryPolicy['max_attempts'])) $rp['max_attempts'] = $retryPolicy['max_attempts'];
            if (isset($retryPolicy['backoff'])) $rp['backoff'] = $retryPolicy['backoff'];
            if (isset($retryPolicy['initial_delay_secs'])) $rp['initial_delay_secs'] = $retryPolicy['initial_delay_secs'];
            if (isset($retryPolicy['max_delay_secs'])) $rp['max_delay_secs'] = $retryPolicy['max_delay_secs'];
            $body['retry_policy'] = $rp;
        }

        $resp = $this->client->request('POST', '/endpoints', $body);
        return $this->mapEndpoint($resp);
    }

    /**
     * Get an endpoint by ID.
     */
    public function get(string $endpointId): array
    {
        $resp = $this->client->request('GET', "/endpoints/{$endpointId}");
        return $this->mapEndpoint($resp);
    }

    /**
     * List all endpoints.
     */
    public function list(): array
    {
        $resp = $this->client->request('GET', '/endpoints');
        return array_map([$this, 'mapEndpoint'], $resp);
    }

    /**
     * Delete an endpoint.
     */
    public function delete(string $endpointId): bool
    {
        $resp = $this->client->request('DELETE', "/endpoints/{$endpointId}");
        return $resp['deleted'] ?? true;
    }

    private function mapEndpoint(array $data): array
    {
        $rp = $data['retry_policy'] ?? null;
        return [
            'id' => $data['id'],
            'url' => $data['url'],
            'description' => $data['description'] ?? null,
            'is_active' => $data['is_active'] ?? false,
            'retry_policy' => $rp ? [
                'max_attempts' => $rp['max_attempts'] ?? null,
                'backoff' => $rp['backoff'] ?? null,
                'initial_delay_secs' => $rp['initial_delay_secs'] ?? null,
                'max_delay_secs' => $rp['max_delay_secs'] ?? null,
            ] : null,
            'created_at' => $data['created_at'] ?? null,
        ];
    }
}

/**
 * Webhooks resource — send, list, replay, batch, and inspect webhooks.
 */
class WebhooksResource
{
    public function __construct(private HookRelayClient $client) {}

    /**
     * Send a webhook.
     */
    public function send(string $endpointId, array $data, ?string $event = null): array
    {
        $body = ['endpoint_id' => $endpointId, 'data' => $data];
        if ($event !== null) $body['event'] = $event;
        $resp = $this->client->request('POST', '/webhooks', $body);
        return $this->mapDelivery($resp);
    }

    /**
     * Get a delivery by ID.
     */
    public function get(string $deliveryId): array
    {
        $resp = $this->client->request('GET', "/webhooks/{$deliveryId}");
        return $this->mapDelivery($resp);
    }

    /**
     * List deliveries with optional filters.
     */
    public function list(?string $status = null, int $page = 1, int $perPage = 20): array
    {
        $params = http_build_query([
            'page' => $page,
            'per_page' => $perPage,
            'status' => $status,
        ]);
        $resp = $this->client->request('GET', "/webhooks?{$params}");
        return [
            'deliveries' => array_map([$this, 'mapDelivery'], $resp['deliveries'] ?? []),
            'total' => $resp['total'] ?? 0,
            'page' => $resp['page'] ?? $page,
            'per_page' => $resp['per_page'] ?? $perPage,
        ];
    }

    /**
     * Replay a delivery.
     */
    public function replay(string $deliveryId): array
    {
        $resp = $this->client->request('POST', "/webhooks/{$deliveryId}/replay");
        return $this->mapDelivery($resp);
    }

    /**
     * Send multiple webhooks in a batch.
     */
    public function batch(array $webhooks): array
    {
        $body = ['webhooks' => array_map(function ($w) {
            $item = ['endpoint_id' => $w['endpoint_id'], 'data' => $w['data']];
            if (isset($w['event'])) $item['event'] = $w['event'];
            return $item;
        }, $webhooks)];

        $resp = $this->client->request('POST', '/webhooks/batch', $body);
        return [
            'deliveries' => array_map([$this, 'mapDelivery'], $resp['deliveries'] ?? []),
            'errors' => $resp['errors'] ?? [],
        ];
    }

    /**
     * Get delivery attempts.
     */
    public function attempts(string $deliveryId): array
    {
        $resp = $this->client->request('GET', "/webhooks/{$deliveryId}/attempts");
        return array_map([$this, 'mapAttempt'], $resp);
    }

    /**
     * Export deliveries.
     */
    public function export(?string $format = null, ?string $status = null,
                           ?string $dateFrom = null, ?string $dateTo = null): mixed
    {
        $params = array_filter([
            'format' => $format,
            'status' => $status,
            'date_from' => $dateFrom,
            'date_to' => $dateTo,
        ], fn($v) => $v !== null);
        $query = http_build_query($params);
        $resp = $this->client->request('GET', "/webhooks/export?{$query}");

        if ($format === 'csv') return $resp;
        return array_map([$this, 'mapDelivery'], $resp);
    }

    private function mapDelivery(array $data): array
    {
        return [
            'id' => $data['id'],
            'endpoint_id' => $data['endpoint_id'] ?? null,
            'event' => $data['event'] ?? null,
            'status' => $data['status'] ?? null,
            'attempt_count' => $data['attempt_count'] ?? 0,
            'response_status' => $data['response_status'] ?? null,
            'replay_count' => $data['replay_count'] ?? 0,
            'created_at' => $data['created_at'] ?? null,
        ];
    }

    private function mapAttempt(array $data): array
    {
        return [
            'id' => $data['id'],
            'attempt_number' => $data['attempt_number'] ?? 0,
            'status_code' => $data['status_code'] ?? null,
            'response_body' => $data['response_body'] ?? null,
            'duration_ms' => $data['duration_ms'] ?? null,
            'error_message' => $data['error_message'] ?? null,
            'created_at' => $data['created_at'] ?? null,
        ];
    }
}

File: sdks/csharp/HookRelayClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

namespace HookRelay
{
    // ==================== Models ====================

    public class HookRelayConfig
    {
        public string ApiKey { get; set; } = "";
        public string BaseUrl { get; set; } = "https://api.hookrelay.io/v1";
        public int Timeout { get; set; } = 30;
    }

    public class RetryPolicy
    {
        [JsonPropertyName("max_attempts")]
        public int? MaxAttempts { get; set; }
        [JsonPropertyName("backoff")]
        public string? Backoff { get; set; }
        [JsonPropertyName("initial_delay_secs")]
        public int? InitialDelaySecs { get; set; }
        [JsonPropertyName("max_delay_secs")]
        public int? MaxDelaySecs { get; set; }
    }

    public class Endpoint
    {
        [JsonPropertyName("id")]
        public string Id { get; set; } = "";
        [JsonPropertyName("url")]
        public string Url { get; set; } = "";
        [JsonPropertyName("description")]
        public string? Description { get; set; }
        [JsonPropertyName("is_active")]
        public bool IsActive { get; set; }
        [JsonPropertyName("retry_policy")]
        public RetryPolicy? RetryPolicy { get; set; }
        [JsonPropertyName("created_at")]
        public string CreatedAt { get; set; } = "";
    }

    public class Delivery
    {
        [JsonPropertyName("id")]
        public string Id { get; set; } = "";
        [JsonPropertyName("endpoint_id")]
        public string EndpointId { get; set; } = "";
        [JsonPropertyName("event")]
        public string? Event { get; set; }
        [JsonPropertyName("status")]
        public string Status { get; set; } = "";
        [JsonPropertyName("attempt_count")]
        public int AttemptCount { get; set; }
        [JsonPropertyName("response_status")]
        public int? ResponseStatus { get; set; }
        [JsonPropertyName("replay_count")]
        public int ReplayCount { get; set; }
        [JsonPropertyName("created_at")]
        public string CreatedAt { get; set; } = "";
    }

    public class DeliveryList
    {
        [JsonPropertyName("deliveries")]
        public List<Delivery> Deliveries { get; set; } = new();
        [JsonPropertyName("total")]
        public int Total { get; set; }
        [JsonPropertyName("page")]
        public int Page { get; set; }
        [JsonPropertyName("per_page")]
        public int PerPage { get; set; }
    }

    public class EndpointList
    {
        [JsonPropertyName("endpoints")]
        public List<Endpoint> Endpoints { get; set; } = new();
        [JsonPropertyName("total")]
        public int Total { get; set; }
        [JsonPropertyName("page")]
        public int Page { get; set; }
        [JsonPropertyName("per_page")]
        public int PerPage { get; set; }
    }

    public class DeliveryAttempt
    {
        [JsonPropertyName("id")]
        public string Id { get; set; } = "";
        [JsonPropertyName("attempt_number")]
        public int AttemptNumber { get; set; }
        [JsonPropertyName("status_code")]
        public int? StatusCode { get; set; }
        [JsonPropertyName("response_body")]
        public string? ResponseBody { get; set; }
        [JsonPropertyName("duration_ms")]
        public long? DurationMs { get; set; }
        [JsonPropertyName("error_message")]
        public string? ErrorMessage { get; set; }
        [JsonPropertyName("created_at")]
        public string CreatedAt { get; set; } = "";
    }

    public class BatchResult
    {
        [JsonPropertyName("deliveries")]
        public List<Delivery> Deliveries { get; set; } = new();
        [JsonPropertyName("errors")]
        public List<JsonElement> Errors { get; set; } = new();
    }

    public class Stats
    {
        [JsonPropertyName("total_deliveries")]
        public int TotalDeliveries { get; set; }
        [JsonPropertyName("delivered")]
        public int Delivered { get; set; }
        [JsonPropertyName("failed")]
        public int Failed { get; set; }
        [JsonPropertyName("pending")]
        public int Pending { get; set; }
        [JsonPropertyName("success_rate")]
        public double SuccessRate { get; set; }
        [JsonPropertyName("endpoints_count")]
        public int EndpointsCount { get; set; }
    }

    // ==================== Exceptions ====================

    public class HookRelayException : Exception
    {
        public int StatusCode { get; }
        public string ErrorCode { get; }

        public HookRelayException(string message, int statusCode = 0, string errorCode = "UNKNOWN")
            : base(message)
        {
            StatusCode = statusCode;
            ErrorCode = errorCode;
        }
    }

    public class AuthenticationException : HookRelayException
    {
        public AuthenticationException(string message = "Unauthorized: invalid or missing API key")
            : base(message, 401, "UNAUTHORIZED") { }
    }

    public class NotFoundException : HookRelayException
    {
        public NotFoundException(string message = "Resource not found")
            : base(message, 404, "NOT_FOUND") { }
    }

    public class RateLimitException : HookRelayException
    {
        public RateLimitException(string message = "Rate limit exceeded")
            : base(message, 429, "RATE_LIMIT_EXCEEDED") { }
    }

    public class ValidationException : HookRelayException
    {
        public ValidationException(string message = "Bad request")
            : base(message, 400, "BAD_REQUEST") { }
    }

    public class PayloadTooLargeException : HookRelayException
    {
        public PayloadTooLargeException(string message = "Payload too large")
            : base(message, 413, "PAYLOAD_TOO_LARGE") { }
    }

    // ==================== Client ====================

    /// <summary>
    /// Official C# client for the HookRelay webhook delivery service.
    /// </summary>
    public class HookRelayClient : IDisposable
    {
        private readonly HttpClient _httpClient;
        private readonly string _apiKey;
        private readonly string _baseUrl;
        private readonly JsonSerializerOptions _jsonOptions;

        public EndpointsResource Endpoints { get; }
        public WebhooksResource Webhooks { get; }

        public HookRelayClient(string apiKey)
            : this(new HookRelayConfig { ApiKey = apiKey }) { }

        public HookRelayClient(HookRelayConfig config)
        {
            _apiKey = config.ApiKey ?? throw new ArgumentNullException(nameof(config.ApiKey));
            _baseUrl = (config.BaseUrl ?? "https://api.hookrelay.io/v1").TrimEnd('/');

            _httpClient = new HttpClient
            {
                Timeout = TimeSpan.FromSeconds(config.Timeout > 0 ? config.Timeout : 30)
            };

            _jsonOptions = new JsonSerializerOptions
            {
                PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
                DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
            };

            Endpoints = new EndpointsResource(this);
            Webhooks = new WebhooksResource(this);
        }

        /// <summary>
        /// Get platform statistics.
        /// </summary>
        public async Task<Stats> GetStatsAsync()
        {
            return await RequestAsync<Stats>("GET", "/stats");
        }

        /// <summary>
        /// Dispose the HTTP client.
        /// </summary>
        public void Dispose()
        {
            _httpClient.Dispose();
        }

        // ==================== Internal HTTP ====================

        internal async Task<T> RequestAsync<T>(string method, string path, object? body = null)
        {
            var response = await SendRequestAsync(method, path, body);
            var content = await response.Content.ReadAsStringAsync();
            return JsonSerializer.Deserialize<T>(content, _jsonOptions)
                   ?? throw new HookRelayException("Failed to deserialize response");
        }

        internal async Task<string> RequestRawAsync(string method, string path, object? body = null)
        {
            var response = await SendRequestAsync(method, path, body);
            return await response.Content.ReadAsStringAsync();
        }

        private async Task<HttpResponseMessage> SendRequestAsync(string method, string path, object? body)
        {
            var url = $"{_baseUrl}{path}";
            var request = new HttpRequestMessage(new HttpMethod(method), url);
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _apiKey);
            request.Headers.Add("User-Agent", "hookrelay-csharp/0.2.0");

            if (body != null && (method == "POST" || method == "PUT" || method == "PATCH"))
            {
                var json = JsonSerializer.Serialize(body, _jsonOptions);
                request.Content = new StringContent(json, Encoding.UTF8, "application/json");
            }

            var response = await _httpClient.SendAsync(request);

            if (response.IsSuccessStatusCode)
            {
                return response;
            }

            var errBody = await response.Content.ReadAsStringAsync();
            var message = $"HTTP {(int)response.StatusCode}";
            try
            {
                var errDoc = JsonDocument.Parse(errBody);
                if (errDoc.RootElement.TryGetProperty("error", out var errObj) &&
                    errObj.TryGetProperty("message", out var msgProp))
                {
                    message = msgProp.GetString() ?? message;
                }
            }
            catch { }

            throw ((int)response.StatusCode) switch
            {
                400 => new ValidationException(message),
                401 => new AuthenticationException(message),
                404 => new NotFoundException(message),
                413 => new PayloadTooLargeException(message),
                429 => new RateLimitException(message),
                _ => new HookRelayException(message, (int)response.StatusCode)
            };
        }
    }

    // ==================== Endpoints Resource ====================

    public class EndpointsResource
    {
        private readonly HookRelayClient _client;

        internal EndpointsResource(HookRelayClient client)
        {
            _client = client;
        }

        /// <summary>
        /// Create a new endpoint.
        /// </summary>
        public async Task<Endpoint> CreateAsync(string url, string? description = null, RetryPolicy? retryPolicy = null)
        {
            var body = new Dictionary<string, object?> { ["url"] = url };
            if (description != null) body["description"] = description;
            if (retryPolicy != null)
            {
                var rp = new Dictionary<string, object?>();
                if (retryPolicy.MaxAttempts.HasValue) rp["max_attempts"] = retryPolicy.MaxAttempts;
                if (retryPolicy.Backoff != null) rp["backoff"] = retryPolicy.Backoff;
                if (retryPolicy.InitialDelaySecs.HasValue) rp["initial_delay_secs"] = retryPolicy.InitialDelaySecs;
                if (retryPolicy.MaxDelaySecs.HasValue) rp["max_delay_secs"] = retryPolicy.MaxDelaySecs;
                body["retry_policy"] = rp;
            }

            return await _client.RequestAsync<Endpoint>("POST", "/endpoints", body);
        }

        /// <summary>
        /// Get an endpoint by ID.
        /// </summary>
        public async Task<Endpoint> GetAsync(string endpointId)
        {
            return await _client.RequestAsync<Endpoint>("GET", $"/endpoints/{endpointId}");
        }

        /// <summary>
        /// List all endpoints with pagination.
        /// </summary>
        public async Task<EndpointList> ListAsync(int page = 1, int perPage = 20)
        {
            return await _client.RequestAsync<EndpointList>("GET", $"/endpoints?page={page}&per_page={perPage}");
        }

        /// <summary>
        /// Delete an endpoint.
        /// </summary>
        public async Task<bool> DeleteAsync(string endpointId)
        {
            var result = await _client.RequestAsync<Dictionary<string, JsonElement>>("DELETE", $"/endpoints/{endpointId}");
            return !result.ContainsKey("deleted") || result["deleted"].GetBoolean();
        }

        /// <summary>
        /// Rotate the signing secret for an endpoint.
        /// </summary>
        public async Task<Dictionary<string, JsonElement>> RotateSecretAsync(string endpointId)
        {
            return await _client.RequestAsync<Dictionary<string, JsonElement>>("POST", $"/endpoints/{endpointId}/rotate-secret");
        }
    }

    // ==================== Webhooks Resource ====================

    public class WebhooksResource
    {
        private readonly HookRelayClient _client;

        internal WebhooksResource(HookRelayClient client)
        {
            _client = client;
        }

        /// <summary>
        /// Send a webhook.
        /// </summary>
        public async Task<Delivery> SendAsync(string endpointId, string? @event = null, Dictionary<string, object>? data = null)
        {
            var body = new Dictionary<string, object?>
            {
                ["endpoint_id"] = endpointId,
                ["data"] = data ?? new Dictionary<string, object>()
            };
            if (@event != null) body["event"] = @event;

            return await _client.RequestAsync<Delivery>("POST", "/webhooks", body);
        }

        /// <summary>
        /// Get a delivery by ID.
        /// </summary>
        public async Task<Delivery> GetAsync(string deliveryId)
        {
            return await _client.RequestAsync<Delivery>("GET", $"/webhooks/{deliveryId}");
        }

        /// <summary>
        /// List deliveries with optional filters.
        /// </summary>
        public async Task<DeliveryList> ListAsync(string? status = null, int page = 1, int perPage = 20)
        {
            var query = $"?page={page}&per_page={perPage}";
            if (status != null) query += $"&status={status}";
            return await _client.RequestAsync<DeliveryList>("GET", $"/webhooks{query}");
        }

        /// <summary>
        /// Replay a delivery.
        /// </summary>
        public async Task<Delivery> ReplayAsync(string deliveryId)
        {
            return await _client.RequestAsync<Delivery>("POST", $"/webhooks/{deliveryId}/replay");
        }

        /// <summary>
        /// Send multiple webhooks in a batch.
        /// </summary>
        public async Task<BatchResult> BatchAsync(List<Dictionary<string, object>> webhooks)
        {
            var body = new Dictionary<string, object> { ["webhooks"] = webhooks };
            return await _client.RequestAsync<BatchResult>("POST", "/webhooks/batch", body);
        }

        /// <summary>
        /// Get delivery attempts.
        /// </summary>
        public async Task<List<DeliveryAttempt>> AttemptsAsync(string deliveryId)
        {
            return await _client.RequestAsync<List<DeliveryAttempt>>("GET", $"/webhooks/{deliveryId}/attempts");
        }

        /// <summary>
        /// Export deliveries.
        /// </summary>
        public async Task<List<Delivery>> ExportAsync(string? format = null, string? status = null,
            string? dateFrom = null, string? dateTo = null)
        {
            var queryParams = new List<string>();
            if (format != null) queryParams.Add($"format={format}");
            if (status != null) queryParams.Add($"status={status}");
            if (dateFrom != null) queryParams.Add($"date_from={dateFrom}");
            if (dateTo != null) queryParams.Add($"date_to={dateTo}");

            var query = queryParams.Count > 0 ? "?" + string.Join("&", queryParams) : "";
            return await _client.RequestAsync<List<Delivery>>("GET", $"/webhooks/export{query}");
        }
    }
}
